cli/open.go

```diff
@@ -0,0 +1,268 @@
+				qp.Add("folder", directory)
+			}
+
+			// We always set the token if we believe we can open without
+			// printing the URI, otherwise the token must be explicitly
+			// requested as it will be printed in plain text.
+			if !insideAWorkspace || generateToken {
+				// Prepare an API key. This is for automagical configuration of
+				// VS Code, however, if running on a local machine we could try
+				// to probe VS Code settings to see if the current configuration
+				// is valid. Future improvement idea.
+				apiKey, err := client.CreateAPIKey(ctx, codersdk.Me)
+				if err != nil {
+					return xerrors.Errorf("create API key: %w", err)
+				}
+				qp.Add("token", apiKey.Key)
+			}
+
+			u.RawQuery = qp.Encode()
+
+			openingPath := workspaceName
+			if directory != "" {
+				openingPath += ":" + directory
+			}
+
+			if insideAWorkspace {
+				_, _ = fmt.Fprintf(inv.Stderr, "Opening %s in %s is not supported inside a workspace, please open the following URI on your local machine instead:\n\n", openingPath, vscodeDesktopName)
+				_, _ = fmt.Fprintf(inv.Stdout, "%s\n", u.String())
+				return nil
+			}
+			_, _ = fmt.Fprintf(inv.Stderr, "Opening %s in %s\n", openingPath, vscodeDesktopName)
+
+			if !testOpenError {
+				err = open.Run(u.String())
+			} else {
+				err = xerrors.New("test.open-error")
+			}
+			if err != nil {
+				if !generateToken {
+					qp.Del("token")
+					u.RawQuery = qp.Encode()
+				}
+
+				_, _ = fmt.Fprintf(inv.Stderr, "Could not automatically open %s in %s: %s\n", openingPath, vscodeDesktopName, err)
+				_, _ = fmt.Fprintf(inv.Stderr, "Please open the following URI instead:\n\n")
+				_, _ = fmt.Fprintf(inv.Stdout, "%s\n", u.String())
+				return nil
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Options = clibase.OptionSet{
+		{
+			Flag: "generate-token",
+			Env:  "CODER_OPEN_VSCODE_GENERATE_TOKEN",
+			Description: fmt.Sprintf(
+				"Generate an auth token and include it in the vscode:// URI. This is for automagical configuration of %s and not needed if already configured. "+
+					"This flag does not need to be specified when running this command on a local machine unless automatic open fails.",
+				vscodeDesktopName,
+			),
+			Value: clibase.BoolOf(&generateToken),
+		},
+		{
+			Flag:        "test.open-error",
+			Description: "Don't run the open command.",
+			Value:       clibase.BoolOf(&testOpenError),
+			Hidden:      true, // This is for testing!
+		},
+	}
+
+	return cmd
+}
+
+// isWindowsAbsPath checks if the path is an absolute path on Windows. On Unix
+// systems the check is very simplistic and does not cover edge cases.
+//
+//nolint:revive // Shadow path variable for readability.
+func isWindowsAbsPath(path string) bool {
+	if runtime.GOOS == "windows" {
+		return filepath.IsAbs(path)
+	}
+
+	switch {
+	case len(path) >= 2 && path[1] == ':':
+		// Path starts with a drive letter.
+		return len(path) == 2 || (len(path) >= 4 && path[2] == '\\' && path[3] == '\\')
+	default:
+		return false
```

Comment:
Maybe I'll need to play around with this some more. I was actually planning on testing with agents on Windows today, but it turns out our win template is broken atm. I know at least the "expanded directory" resolves to `C:\...` via test runners. I suspected `/` might work in VS Code but out of precaution I wanted to transform them.

Perhaps we can just make everything slash and only have the logic for detecting abs path on Windows.

Reply:
I added some handling for the cases you mentioned, and relaxed the abs requirement to include `/`, `\`. I know there are more exceptions like `\\.UNC\` paths and the like, but I don't know if it's worth adding preemptive handling for that.

Reply:
Actually, I can't seem to make CI happy because Go `filepath.IsAbs` on windows seems to think differently? https://github.com/coder/coder/actions/runs/7262677715/job/19786347550?pr=11191#step:5:496